app/screens/edit_profile/edit_profile.tsx

```diff
@@ -152,98 +154,148 @@ const EditProfile = ({
         loadCustomAttributes();
     }, [currentUser, serverUrl, enableCustomAttributes, customAttributesSet]);
 
+    const fieldLockConfig = useMemo(() => {
+        const service = currentUser?.authService || '';
+        const includesSsoService = (sso: string) => ['gitlab', 'google', 'office365'].includes(sso);
+        const isSAMLOrLDAP = (protocol: string) => ['ldap', 'saml'].includes(protocol);
+
+        const config = {
+            email: true, // Always disabled in form
+            firstName: (isSAMLOrLDAP(service) && lockedFirstName) || includesSsoService(service),
+            lastName: (isSAMLOrLDAP(service) && lockedLastName) || includesSsoService(service),
+            nickname: isSAMLOrLDAP(service) && lockedNickname,
+            position: isSAMLOrLDAP(service) && lockedPosition,
+            username: service !== '',
+        };
+
+        return config;
+    }, [currentUser?.authService, lockedFirstName, lockedLastName, lockedNickname, lockedPosition]);
+
+    const updateUserInfo = useCallback((updates: Partial<UserInfo>, key: string, newValue: string, oldValue: string): Partial<UserInfo> => {
+        const val = newValue.trim();
+        const isLocked = fieldLockConfig[key as keyof typeof fieldLockConfig] || false;
+        const hasChanged = val !== oldValue;
+
+        if (!isLocked && hasChanged) {
+            updates[key as keyof UserInfo] = val as any;
+        }
+        return updates;
+    }, [fieldLockConfig]);
```

Comment:
If you pass the fieldLockConfig as an argument, you can remove this dependency, which allows you to move this out of the component. Doing so, you can also get `buildUserInfoUpdates` also outside of the component. That should remove at least one dependency on submitUsers.